### PR TITLE
Implements abort and client disconnect

### PR DESCRIPTION
**Pull request by jkotalik**
For #461 and #438

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 1]
```diff
@@ -96,15 +96,31 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS OnAsyncCompletion(IntPt
             return NativeMethods.REQUEST_NOTIFICATION_STATUS.RQ_NOTIFICATION_PENDING;
         }
 
-        private static void CompleteRequest(IISHttpContext context)
+        private static bool OnClientDisconnect(IntPtr pvManagedHttpContext)
+        {
+            var context = (IISHttpContext)GCHandle.FromIntPtr(pvManagedHttpContext).Target;
+            // There is a race here between the context cleaning up and notifying the cancellation token.
+            //
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: I’d like to understand this comment

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs [hunk 2]
```diff
@@ -19,10 +19,10 @@ public IISHttpContextOfT(PipeFactory pipeFactory, IHttpApplication<TContext> app
             _application = application;
         }
 
-        public override async Task ProcessRequestAsync()
+        public override async Task<bool> ProcessRequestAsync()
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs]: Is this change related? Why isn’t this another PR?

Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs]: It is related. If the request is aborted, we would like to return FINISH_REQUEST to the IIS pipeline. Unless you want to go with the logic of always returning continue.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 1 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: I'll clean that up. My b.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs (hunk 2 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs]: Is that what’s happening here? I don’t see it. Where does abort cause success to be false?

Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs]: Calling abort could throw an exception from user code on writing (or other operations on the context).

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs]: What does that have to do with ending the request? The user code can catch the exception.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 3]
```diff
@@ -96,7 +96,15 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS OnAsyncCompletion(IntPt
...
         }
 
-        private static void CompleteRequest(IISHttpContext context)
+        private static bool OnClientDisconnect(IntPtr pvManagedHttpContext)
+        {
+            var context = (IISHttpContext)GCHandle.FromIntPtr(pvManagedHttpContext).Target;
+            context?.OnClientDisconnect();
+            return true;
+        }
+
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: nit: extra space

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs (hunk 2 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContextOfT.cs]: https://github.com/aspnet/IISIntegration/pull/469

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 3 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: I removed that, not sure why it is still showing up.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs [hunk 4]
```diff
@@ -938,6 +1003,11 @@ internal void OnAsyncCompletion(int hr, int cbBytes)
             }
         }
 
+        internal void OnClientDisconnect()
+        {
+            Task.Run(() => CancelRequestAbortedToken());
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs]: Ahh closure. Fix that 😄

**Comment by jkotalik** [comment]
Pending https://github.com/aspnet/AspNetCoreModule/pull/227, which has a few issues that need to be addressed.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs [hunk 5]
```diff
@@ -159,6 +160,69 @@ internal unsafe IISHttpContext(PipeFactory pipeFactory, IntPtr pHttpContext)
...
+                return cts;
+            }
+        }
+
+        private void CancelRequestAbortedToken()
+        {
+            try
+            {
+                RequestAbortedSource.Cancel();
+                _abortedCts = null;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs]: No reason to null this out. It risks race conditions and null refs

**Comment by jkotalik** [comment]
Will reopen once shim is in.
